### PR TITLE
Realize fast-forwarding of the current branch by `git merge --ff-only` instead of `git fetch`

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+          /* important */ true,
+          /* error */ null);
+      gitRepository.update();
+    }
+  }
+
+  private void showUpdates(VirtualFile root, GitRevisionNumber currentRev, Label beforeLabel) {
+    try {
+      UpdatedFiles files = UpdatedFiles.create();
+      @SuppressWarnings("UnstableApiUsage") var collector = new MergeChangeCollector(project, root, currentRev);
```
Comment (micpiotrowski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:190]: `MergeChangeCollector`:
```
@Deprecated
@ApiStatus.ScheduledForRemoval(inVersion = "2020.3")
```
:confused:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        Notification notification;
+        if (notificationData != null) {
+          String title = getTitleForUpdateNotification(notificationData.getUpdatedFilesCount(),
+              notificationData.getReceivedCommitsCount());
+          String content = getBodyForUpdateNotification(notificationData.getFilteredCommitsCount());
+          notification = VcsNotifier.STANDARD_NOTIFICATION.createNotification(title,
+              content,
+              INFORMATION,
+              /* listener */ null);
+          notification.addAction(NotificationAction.createSimple(getString(
```
Comment (micpiotrowski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:145]: Why not just `VcsNotifier.getInstance(project).notifyInfo(...)` here?

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+              "action.GitMachete.GitCommandUpdatingCurrentBranchBackgroundable.notification.message.view-commits"),
+              notificationData.getViewCommitAction()));
+
+        } else {
+          // When the pull results with no commits, there is no git update info (as log).
+          // Based on that we know that all files are up-to-date.
+          notification = VcsNotifier.STANDARD_NOTIFICATION.createNotification(
```
Comment (micpiotrowski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:152]: Ditto here?

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/MergeCurrentBranchFastForwardOnlyBackgroundable.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      GitRepository gitRepository,
+      String taskTitle,
+      String targetBranchName) {
+    super(project, gitRepository, taskTitle);
+    this.targetBranchName = targetBranchName;
+  }
+
+  @Override
+  protected String getOperationName() {
+    return "Fast-forward merge";
```
Comment (micpiotrowski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/MergeCurrentBranchFastForwardOnlyBackgroundable.java:26]: I18n => move to bundle (here and in other places)?

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java (hunk 1 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:190]: Whoops @mkondratek do we have a remedy for this one? :confused:

Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:190]: We can simply replace the `VirtualFile root` with `GitRepository repository` (the suggested constructor)

image

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java (hunk 2 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:145]: Hmm good question... @mkondratek :question:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java (hunk 1 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:190]: Whoops asked before checking first, sorry :xd:

Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:190]: Done, fixed

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/MergeCurrentBranchFastForwardOnlyBackgroundable.java (hunk 4 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/MergeCurrentBranchFastForwardOnlyBackgroundable.java:26]: Done both here and in `PullCurrentBranchFastForwardOnlyBackgroundable`

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java (hunk 3 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:152]: (@mkondratek)

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java (hunk 1 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java]: Whoops not so easy, the newer c'tor isn't available on 2020.1 yet... let me do some reflection magic to provide smooth transition to 2020.3

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+          format(getString("action.GitMachete.GitCommandUpdatingCurrentBranchBackgroundable.notification.fail"),
+              getOperationName()),
+          result.getErrorOutputAsJoinedString(),
+          /* important */ true,
+          /* error */ null);
+      gitRepository.update();
+    }
+  }
+
+  // TODO (#496): replace with a non-reflective constructor call
```
Comment (mkondratek) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:189]: awwww

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java (hunk 3 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:154]: you are right, here we CAN use the regular `VcsNotifier.getInstance(project).notifyInfo(...)`

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java (hunk 2 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:147]: I took a cue from `git4idea.actions.GitMergeAction#handleResult`

@micpiotrowski how would you provide the `View commits` action to the notification using `VcsNotifier.getInstance(project).notifyInfo(...)`?

Comment (micpiotrowski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:147]: Yeah, I didn't notice that. My fault.

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java (hunk 3 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/GitCommandUpdatingCurrentBranchBackgroundable.java:154]: Hmm given that the other place must use `STANDARD_NOTIFICATION.createNotification`... I'd stick to the same method of creating a notification in both places

**Pull request merged**